File: sync_pr_context.py
#!/usr/bin/env -S uv run --quiet --script
# /// script
# requires-python = ">=3.12"
# dependencies = []
# ///

"""
Database Sync CLI - Sync PR context from GitHub to SQLite database

This script reads config.yaml, scans for PRs with for-landing/for-review labels,
and saves their complete context to the database for offline agent testing.

Usage:
    ./sync_pr_context.py [--config config.yaml] [--repo REPO_NAME] [--pr PR_NUMBER]

Examples:
    # Sync all PRs from all repos in config
    ./sync_pr_context.py

    # Sync specific repo
    ./sync_pr_context.py --repo merge-god

    # Sync specific PR
    ./sync_pr_context.py --repo merge-god --pr 134

    # Use different config file
    ./sync_pr_context.py --config /path/to/config.yaml
"""

import argparse
import json
import sys
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

import yaml

from db_operations import StateDatabase
from github_ops import gather_pr_context
from git_ops import get_repo_prs_with_labels, get_default_branch


def log_json(event_type: str, data: dict[str, Any]) -> None:
    """Emit structured JSON logs with timestamp"""
    log_entry = {
        "timestamp": datetime.now(timezone.utc).isoformat().replace("+00:00", "Z"),
        "event": event_type,
        "data": data,
    }
    print(json.dumps(log_entry), flush=True)


def load_config(config_path: Path) -> dict[str, Any]:
    """Load configuration from YAML file"""
    if not config_path.exists():
        raise FileNotFoundError(f"Config file not found: {config_path}")

    with open(config_path) as f:
        config = yaml.safe_load(f)

    if not config or "repos" not in config:
        raise ValueError(f"Invalid config file: missing 'repos' section")

    return config


def sync_pr_to_database(
    db: StateDatabase,
    repo_path: Path,
    repo_name: str,
    pr_number: int
) -> bool:
    """
    Sync a single PR's context to the database.

    Args:
        db: Database instance
        repo_path: Path to git repository
        repo_name: Repository name
        pr_number: PR number

    Returns:
        True if sync successful, False otherwise
    """
    log_json("sync_pr", {
        "action": "start",
        "repo": repo_name,
        "pr_number": pr_number
    })

    try:
        # Gather PR context using existing functions
        pr_details, pr_context = gather_pr_context(
            repo_path=repo_path,
            pr_number=pr_number,
            mode="for-landing"  # Default mode, doesn't matter for context gathering
        )

        if not pr_details or not pr_context:
            log_json("sync_pr", {
                "action": "error",
                "repo": repo_name,
                "pr_number": pr_number,
                "error": "Failed to gather PR context"
            })
            return False

        # Save to database
        db.save_pr_context(
            repo_name=repo_name,
            pr_number=pr_number,
            pr_details=pr_details,
            pr_context=pr_context
        )

        # Also save PR snapshot
        pr_snapshot = {
            "number": pr_number,
            "title": pr_details.get("title", ""),
            "state": "open",
            "head_branch": pr_details.get("headRefName", ""),
            "base_branch": pr_details.get("baseRefName", "main"),
            "author": pr_details.get("author", {}).get("login", "unknown"),
            "draft": False,
            "ci_status": "unknown",
            "labels": pr_details.get("labels", []),
            "created_at": datetime.now(timezone.utc),
            "updated_at": datetime.now(timezone.utc)
        }
        db.save_pr_snapshot(repo_name, pr_snapshot)

        log_json("sync_pr", {
            "action": "complete",
            "repo": repo_name,
            "pr_number": pr_number,
            "diff_size": len(pr_context.get("diff", "")),
            "comment_count": len(pr_context.get("comments", [])),
            "review_comment_count": len(pr_context.get("review_comments", [])),
            "file_count": len(pr_context.get("files", []))
        })

        return True

    except Exception as e:
        log_json("sync_pr", {
            "action": "error",
            "repo": repo_name,
            "pr_number": pr_number,
            "error": str(e),
            "error_type": type(e).__name__
        })
        return False


def sync_repo(
    db: StateDatabase,
    repo_config: dict[str, Any],
    specific_pr: int | None = None
) -> dict[str, Any]:
    """
    Sync all PRs (or specific PR) from a repository.

    Args:
        db: Database instance
        repo_config: Repository configuration from config.yaml
        specific_pr: Optional specific PR number to sync

    Returns:
        Statistics about the sync operation
    """
    repo_path = Path(repo_config["path"])
    repo_name = repo_config.get("name", repo_path.name)

    if not repo_path.exists():
        log_json("sync_repo", {
            "action": "error",
            "repo": repo_name,
            "error": f"Repository path does not exist: {repo_path}"
        })
        return {"success": False, "error": "path_not_found"}

    log_json("sync_repo", {
        "action": "start",
        "repo": repo_name,
        "path": str(repo_path),
        "specific_pr": specific_pr
    })

    stats = {
        "repo": repo_name,
        "total": 0,
        "succeeded": 0,
        "failed": 0,
        "prs": []
    }

    try:
        if specific_pr:
            # Sync specific PR
            stats["total"] = 1
            success = sync_pr_to_database(db, repo_path, repo_name, specific_pr)
            if success:
                stats["succeeded"] = 1
                stats["prs"].append(specific_pr)
            else:
                stats["failed"] = 1
        else:
            # Discover PRs with for-landing or for-review labels
            default_branch = get_default_branch(repo_path)
            prs_for_landing = get_repo_prs_with_labels(repo_path, ["for-landing"])
            prs_for_review = get_repo_prs_with_labels(repo_path, ["for-review"])

            all_prs = set(prs_for_landing + prs_for_review)
            stats["total"] = len(all_prs)

            log_json("sync_repo", {
                "action": "discovered_prs",
                "repo": repo_name,
                "pr_count": len(all_prs),
                "pr_numbers": sorted(list(all_prs))
            })

            if not all_prs:
                log_json("sync_repo", {
                    "action": "warning",
                    "repo": repo_name,
                    "warning": "No PRs found with for-landing or for-review labels"
                })

            # Sync each PR
            for pr_number in sorted(all_prs):
                success = sync_pr_to_database(db, repo_path, repo_name, pr_number)
                if success:
                    stats["succeeded"] += 1
                    stats["prs"].append(pr_number)
                else:
                    stats["failed"] += 1

        stats["success"] = stats["failed"] == 0

        log_json("sync_repo", {
            "action": "complete",
            "repo": repo_name,
            "stats": stats
        })

        return stats

    except Exception as e:
        log_json("sync_repo", {
            "action": "error",
            "repo": repo_name,
            "error": str(e),
            "error_type": type(e).__name__
        })
        return {
            **stats,
            "success": False,
            "error": str(e)
        }


def main() -> int:
    """Main entry point"""
    parser = argparse.ArgumentParser(
        description="Sync PR context from GitHub to SQLite database",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
  # Sync all PRs from all repos in config
  ./sync_pr_context.py

  # Sync specific repo
  ./sync_pr_context.py --repo merge-god

  # Sync specific PR
  ./sync_pr_context.py --repo merge-god --pr 134

  # Use different config file
  ./sync_pr_context.py --config /path/to/config.yaml

This script enables database caching for offline agent testing and debugging.
It gathers complete PR context (diff, comments, CI status) and stores it in
SQLite for use with run_agent_from_db.py.
        """
    )

    parser.add_argument(
        "--config",
        type=Path,
        default=Path("config.yaml"),
        help="Path to config.yaml file (default: config.yaml)"
    )

    parser.add_argument(
        "--db",
        type=Path,
        default=Path("merge-god-state.db"),
        help="Path to SQLite database (default: merge-god-state.db)"
    )

    parser.add_argument(
        "--repo",
        type=str,
        help="Sync specific repository by name"
    )

    parser.add_argument(
        "--pr",
        type=int,
        help="Sync specific PR number (requires --repo)"
    )

    args = parser.parse_args()

    # Validate arguments
    if args.pr and not args.repo:
        log_json("error", {
            "error": "--pr requires --repo to be specified"
        })
        return 1

    # Load configuration
    try:
        config = load_config(args.config)
    except Exception as e:
        log_json("error", {
            "error": f"Failed to load config: {e}",
            "config_path": str(args.config)
        })
        return 1

    # Initialize database
    try:
        db = StateDatabase(args.db)
    except Exception as e:
        log_json("error", {
            "error": f"Failed to initialize database: {e}",
            "db_path": str(args.db)
        })
        return 1

    log_json("sync", {
        "action": "start",
        "config": str(args.config),
        "database": str(args.db),
        "repo_filter": args.repo,
        "pr_filter": args.pr
    })

    # Sync repositories
    all_stats = []
    for repo_config in config["repos"]:
        # Skip if not enabled
        if not repo_config.get("enabled", True):
            log_json("sync", {
                "action": "skip",
                "repo": repo_config.get("name", repo_config["path"]),
                "reason": "disabled in config"
            })
            continue

        # Skip if filtering by repo and this isn't it
        if args.repo and repo_config.get("name") != args.repo:
            continue

        stats = sync_repo(db, repo_config, specific_pr=args.pr)
        all_stats.append(stats)

    # Summary
    total_prs = sum(s["total"] for s in all_stats)
    succeeded = sum(s["succeeded"] for s in all_stats)
    failed = sum(s["failed"] for s in all_stats)

    log_json("sync", {
        "action": "complete",
        "total_prs": total_prs,
        "succeeded": succeeded,
        "failed": failed,
        "success_rate": round(succeeded / total_prs * 100, 1) if total_prs > 0 else 0
    })

    return 0 if failed == 0 else 1


if __name__ == "__main__":
    try:
        sys.exit(main())
    except KeyboardInterrupt:
        log_json("shutdown", {"reason": "keyboard_interrupt"})
        sys.exit(130)
    except Exception as e:
        log_json("fatal_error", {
            "error": str(e),
            "error_type": type(e).__name__
        })
        sys.exit(1)

File: sync_pr_context.ts
#!/usr/bin/env node
/**
 * Database Sync CLI - Sync PR context from GitHub to SQLite database.
 *
 * Ported from sync_pr_context.py. Reads config.yaml, scans for PRs with
 * for-landing/for-review labels, and saves their complete context to the
 * database for offline agent testing.
 *
 * Usage:
 *     ./sync_pr_context.ts [--config config.yaml] [--repo REPO_NAME] [--pr PR_NUMBER]
 */

import { chdir } from "node:process";
import { existsSync, readFileSync } from "node:fs";
import { basename, resolve } from "node:path";
import { parseArgs } from "node:util";
import { pathToFileURL } from "node:url";

import YAML from "yaml";

import { SyncStore, GitClient, createPullRequest, PRState } from "@merge-god/github-sync";
import { gather_pr_context, getOpenPrs, getPrDetails } from "./pr-loop";

interface RepoConfig {
  path?: string;
  name?: string;
  enabled?: boolean;
}

interface SyncStats {
  repo: string;
  total: number;
  succeeded: number;
  failed: number;
  prs: number[];
  success?: boolean;
  error?: string;
}

function errMsg(e: unknown): string {
  return e instanceof Error ? e.message : String(e);
}

function typeName(e: unknown): string {
  return e instanceof Error ? e.constructor.name : typeof e;
}

function logJson(eventType: string, data: Record<string, unknown>): void {
  const entry = {
    timestamp: new Date().toISOString().replace("+00:00", "Z"),
    event: eventType,
    data,
  };
  console.log(JSON.stringify(entry));
}

function loadConfig(configPath: string): Record<string, unknown> {
  if (!existsSync(configPath)) {
    throw new Error(`Config file not found: ${configPath}`);
  }
  const text = readFileSync(configPath, "utf8");
  const config = YAML.parse(text);
  if (!config || typeof config !== "object" || Array.isArray(config)) {
    throw new Error("Invalid config file: expected a mapping at the top level");
  }
  if (!("repos" in (config as Record<string, unknown>))) {
    throw new Error("Invalid config file: missing 'repos' section");
  }
  return config as Record<string, unknown>;
}

function authorLogin(details: Record<string, unknown>): string {
  const author = details["author"];
  if (typeof author === "object" && author !== null) {
    const login = (author as Record<string, unknown>)["login"];
    if (typeof login === "string") return login;
  }
  return "unknown";
}

async function syncPrToDatabase(
  db: SyncStore,
  repoPath: string,
  repoName: string,
  prNumber: number,
): Promise<boolean> {
  logJson("sync_pr", { action: "start", repo: repoName, pr_number: prNumber });

  try {
    chdir(repoPath);

    const details = getPrDetails(prNumber);
    const headBranch = (details["headRefName"] as string | undefined) ?? "";
    const baseBranch = (details["baseRefName"] as string | undefined) ?? "main";
    const url = (details["url"] as string | undefined) ?? "";

    const [prDetails, prContext] = await gather_pr_context(
      prNumber,
      headBranch,
      baseBranch,
      url,
    );

    if (!prDetails || Object.keys(prDetails).length === 0) {
      logJson("sync_pr", {
        action: "error",
        repo: repoName,
        pr_number: prNumber,
        error: "Failed to gather PR context",
      });
      return false;
    }

    await db.savePrContext(repoName, prNumber, prDetails, prContext);

    const pr = createPullRequest({
      number: prNumber,
      title: (prDetails["title"] as string | undefined) ?? "",
      state: PRState.OPEN,
      head_branch: (prDetails["headRefName"] as string | undefined) ?? "",
      base_branch: (prDetails["baseRefName"] as string | undefined) ?? "main",
      author: authorLogin(prDetails),
      url,
      created_at: new Date(),
      updated_at: new Date(),
      draft: false,
      labels: ((prDetails["labels"] as unknown[] | undefined) ?? []) as string[],
    });
    await db.savePrSnapshot(repoName, pr);

    const diff = (prContext["diff"] as string | undefined) ?? "";
    const comments = (prContext["comments"] as unknown[] | undefined) ?? [];
    const reviewComments = (prContext["review_comments"] as unknown[] | undefined) ?? [];
    const files = (prContext["files"] as unknown[] | undefined) ?? [];

    logJson("sync_pr", {
      action: "complete",
      repo: repoName,
      pr_number: prNumber,
      diff_size: diff.length,
      comment_count: comments.length,
      review_comment_count: reviewComments.length,
      file_count: files.length,
    });

    return true;
  } catch (e) {
    logJson("sync_pr", {
      action: "error",
      repo: repoName,
      pr_number: prNumber,
      error: errMsg(e),
      error_type: typeName(e),
    });
    return false;
  }
}

async function syncRepo(
  db: SyncStore,
  repoConfig: RepoConfig,
  specificPr?: number,
): Promise<SyncStats> {
  const repoPathStr = repoConfig.path;
  if (!repoPathStr) {
    const fallbackName = repoConfig.name ?? "<unknown>";
    logJson("sync_repo", {
      action: "error",
      repo: fallbackName,
      error: "Repository config missing 'path'",
    });
    return {
      repo: fallbackName,
      total: 0,
      succeeded: 0,
      failed: 0,
      prs: [],
      success: false,
      error: "missing_path",
    };
  }

  const repoPath = resolve(repoPathStr);
  const repoName = repoConfig.name ?? basename(repoPath);

  if (!existsSync(repoPath)) {
    logJson("sync_repo", {
      action: "error",
      repo: repoName,
      error: `Repository path does not exist: ${repoPath}`,
    });
    return {
      repo: repoName,
      total: 0,
      succeeded: 0,
      failed: 0,
      prs: [],
      success: false,
      error: "path_not_found",
    };
  }

  logJson("sync_repo", {
    action: "start",
    repo: repoName,
    path: repoPath,
    specific_pr: specificPr ?? null,
  });

  const stats: SyncStats = {
    repo: repoName,
    total: 0,
    succeeded: 0,
    failed: 0,
    prs: [],
  };

  try {
    chdir(repoPath);

    if (specificPr !== undefined) {
      stats.total = 1;
      const success = await syncPrToDatabase(db, repoPath, repoName, specificPr);
      if (success) {
        stats.succeeded = 1;
        stats.prs.push(specificPr);
      } else {
        stats.failed = 1;
      }
    } else {
      await new GitClient(repoPath).getDefaultBranch();

      const categorized = getOpenPrs();
      const landing = (categorized["for-landing"] ?? []).map(
        (pr) => pr["number"] as number,
      );
      const review = (categorized["for-review"] ?? []).map(
        (pr) => pr["number"] as number,
      );
      const allPrs = new Set<number>([...landing, ...review]);
      stats.total = allPrs.size;

      const sortedPrs = [...allPrs].sort((a, b) => a - b);

      logJson("sync_repo", {
        action: "discovered_prs",
        repo: repoName,
        pr_count: allPrs.size,
        pr_numbers: sortedPrs,
      });

      if (allPrs.size === 0) {
        logJson("sync_repo", {
          action: "warning",
          repo: repoName,
          warning: "No PRs found with for-landing or for-review labels",
        });
      }

      for (const prNumber of sortedPrs) {
        const success = await syncPrToDatabase(db, repoPath, repoName, prNumber);
        if (success) {
          stats.succeeded += 1;
          stats.prs.push(prNumber);
        } else {
          stats.failed += 1;
        }
      }
    }

    stats.success = stats.failed === 0;

    logJson("sync_repo", { action: "complete", repo: repoName, stats });

    return stats;
  } catch (e) {
    logJson("sync_repo", {
      action: "error",
      repo: repoName,
      error: errMsg(e),
      error_type: typeName(e),
    });
    return { ...stats, success: false, error: errMsg(e) };
  }
}

async function main(): Promise<number> {
  const { values } = parseArgs({
    options: {
      config: { type: "string", default: "config.yaml" },
      db: { type: "string", default: "merge-god-state.db" },
      repo: { type: "string" },
      pr: { type: "string" },
    },
    strict: true,
    allowPositionals: false,
  });

  let prNumber: number | undefined;
  if (values.pr !== undefined) {
    prNumber = Number.parseInt(values.pr, 10);
    if (Number.isNaN(prNumber)) {
      logJson("error", { error: `Invalid --pr value: ${values.pr}` });
      return 1;
    }
  }

  if (prNumber !== undefined && !values.repo) {
    logJson("error", { error: "--pr requires --repo to be specified" });
    return 1;
  }

  const configPath = values.config ?? "config.yaml";
  let config: Record<string, unknown>;
  try {
    config = loadConfig(configPath);
  } catch (e) {
    logJson("error", {
      error: `Failed to load config: ${errMsg(e)}`,
      config_path: configPath,
    });
    return 1;
  }

  const dbPath = values.db ?? "merge-god-state.db";
  const db = new SyncStore(dbPath);
  try {
    await db.initialize();
  } catch (e) {
    logJson("error", {
      error: `Failed to initialize database: ${errMsg(e)}`,
      db_path: dbPath,
    });
    await db.close();
    return 1;
  }

  logJson("sync", {
    action: "start",
    config: configPath,
    database: dbPath,
    repo_filter: values.repo ?? null,
    pr_filter: prNumber ?? null,
  });

  const repos = config["repos"];
  if (!Array.isArray(repos)) {
    logJson("error", { error: "Config 'repos' section is not a list" });
    return 1;
  }

  const allStats: SyncStats[] = [];
  for (const repoRaw of repos) {
    const repoConfig = repoRaw as RepoConfig;

    if (!(repoConfig.enabled ?? true)) {
      logJson("sync", {
        action: "skip",
        repo: repoConfig.name ?? repoConfig.path,
        reason: "disabled in config",
      });
      continue;
    }

    if (values.repo && repoConfig.name !== values.repo) {
      continue;
    }

    const stats = await syncRepo(db, repoConfig, prNumber);
    allStats.push(stats);
  }

  const totalPrs = allStats.reduce((sum, s) => sum + s.total, 0);
  const succeeded = allStats.reduce((sum, s) => sum + s.succeeded, 0);
  const failed = allStats.reduce((sum, s) => sum + s.failed, 0);

  logJson("sync", {
    action: "complete",
    total_prs: totalPrs,
    succeeded,
    failed,
    success_rate: totalPrs > 0 ? Math.round((succeeded / totalPrs) * 1000) / 10 : 0,
  });

  await db.close();

  return failed === 0 ? 0 : 1;
}

if (import.meta.url === pathToFileURL(process.argv[1] ?? "").href) {
  process.on("SIGINT", () => {
    logJson("shutdown", { reason: "keyboard_interrupt" });
    process.exit(130);
  });
  main()
    .then((code) => process.exit(code))
    .catch((e) => {
      logJson("fatal_error", { error: errMsg(e), error_type: typeName(e) });
      process.exit(1);
    });
}
